play.data.Form uses DI now via FormFactory

mkurz opened the pull request:
Fixes #5678

Introduces a `FormBuilder` which can be injected to create a `Form` with all the dependency it needs.
I completely got rid of any static `Form.form` reference in the codebase.
@gmethvin Is that the solution that you have in mind?

If this PR makes it in it makes it **much** more easier to fix/implement #5666/#5670 and #5619 and #5620.

mkurz commented:
CI not starting here as well. I waited 10 minutes. Nothing. That's what it looks like:
![validation_not_starting](https://cloud.githubusercontent.com/assets/644927/13252571/c77646c8-da37-11e5-95aa-090503fc5fe7.png)
Also checked https://travis-ci.org/playframework/playframework: There is absolutely **no** activity. Looks like it didn't get triggered.
/cc @jroper maybe you are interested.

mkurz commented:
After closing/reopening it worked...

mkurz commented:
Also this PR basically reverts #5671.

mkurz commented:
Anyway - build is green

gmethvin commented:
This looks good. Do you have any opinion on the name? Not sure if we should call it `FormBuilder` or `FormCreator` or `FormHelper` or something else. It's not a builder in the traditional sense so maybe we should use a different name? I don't have a strong opinion either way.

mkurz commented:
Hmmm... Right now I would vote for `FormCreator`, because that's what it actually does - it creates Forms :smile: I like the name. But other opinions are welcome.
I can change the name tomorrow. Good night!

gmethvin commented on documentation/manual/working/javaGuide/main/forms/code/javaguide/forms/JavaFormHelpers.scala:
```diff
@@ -12,36 +13,37 @@ object JavaFormHelpers extends PlaySpecification {
-      u.setEmails(util.Arrays.asList("a@a", "b@b"))
-      val userForm = Form.form(classOf[UserForm]).fill(u)
-      def segment(name: String) = {
+      def segment(name: String, app: Application) = {
```
You can just make it
```scala
def segment(name: String)(implicit app: Application)
```
so the function call in the tests stays the same

gmethvin commented:
Maybe we can also add the accessor for the backed type to fix #5354

gmethvin commented:
`FormFactory` is another possible (and accurate) name.

mkurz commented:
Done.
* Renamed to `FormFactory`.
* Added `getBackedType` and made `blankInstance` `protected`. Fixes #5354
* Use `(implicit app: Application)`, like you suggested.

Ready to merge.

gmethvin commented:
:+1:

mkurz commented:
Awesome!
